qmule/test/org/qcmg/qmule/SamToFastqWithHeadersTest.java

```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Test
+    public void getFastqFromSamReverse() {
+        SAMRecord record = new SAMRecord(null);
+        record.setReadName("ERR194147.421820475");
+
+        // first read
+        record.setReadBases("TTCTTTGGCCTAATGACATGGCTATTAGTGCACAAGGAAATGGTCAAAAATGGGAAGAAATGTAGGTCACAAAATATTGCACAAAGCTATACTTACTT".getBytes());
+        record.setBaseQualityString("??????????????????????????????????????????????????????????????????????????????????????????????????");
+        record.setFlags(83);
+        record.setMappingQuality(60);
+        record.setCigarString("98M");
+        record.setInferredInsertSize(-323);
+        record.setAttribute("OQ", "??????????????????????????????????????????????????????????????????????????????????????????????????");
+        record.setAttribute("MQ", (Integer) 60);
+        record.setAttribute("MC", "3S98M");
+        record.setAttribute("ZH", "/1");
+        record.setAttribute("ZT", "CTG+???");
+
+        FastqRecord fq = SamToFastqWithHeaders.getFastqRecordFromSamRecord(record, 1, 0, null, null, 0, null, true, null);
+        assertEquals("ERR194147.421820475/1", fq.getReadName());
+        assertEquals("AAGTAAGTATAGCTTTGTGCAATATTTTGTGACCTACATTTCTTCCCATTTTTGACCATTTCCTTGTGCACTAATAGCCATGTCATTAGGCCAAAGAACTG", fq.getReadString());
+        assertEquals("?????????????????????????????????????????????????????????????????????????????????????????????????????", fq.getBaseQualityString());
+    }
+    @Test
+    public void getFastqFromSamDodgyTag() {
+        SAMRecord record = new SAMRecord(null);
+        record.setReadName("ERR194147.1758538");
+
+        // first read
+        record.setReadBases("AAATGAGGGAAGAAAAGAGTTAAATGCATGTTGATTCCAAGCCCCCGCCTGCCGGGGGGACAGCGGGAGGTTGGAGCACGCAGCCCTGGTGCCTGGT".getBytes());
+        record.setBaseQualityString("???????????????????5???????????????????????????????????????&5?????????+??55??????????????????????");
+        record.setFlags(99);
+        record.setMappingQuality(60);
+        record.setCigarString("97M");
+        record.setInferredInsertSize(330);
+        record.setAttribute("OQ", "???????????????????5???????????????????????????????????????&5?????????+??55??????????????????????");
+        record.setAttribute("MQ", (Integer) 60);
+        record.setAttribute("MC", "3S98M");
+        record.setAttribute("ZH", "/1");
+        record.setAttribute("ZT", "GCGA+???'");
+
+        FastqRecord fq = SamToFastqWithHeaders.getFastqRecordFromSamRecord(record, 1, 0, null, null, 0, null, true, null);
+        assertEquals("ERR194147.1758538/1", fq.getReadName());
+        assertEquals("AAATGAGGGAAGAAAAGAGTTAAATGCATGTTGATTCCAAGCCCCCGCCTGCCGGGGGGACAGCGGGAGGTTGGAGCACGCAGCCCTGGTGCCTGGT" + "GCGA", fq.getReadString());
+        assertEquals("???????????????????5???????????????????????????????????????&5?????????+??55?????????????????????????'", fq.getBaseQualityString());
+    }
+}
```

Comment:
Do we need to demonstrate reconstructing a read id with a space in it like `@ERR194147.1758538/1 ACGTACGT` ? I mean it's pretty obvious that it works no matter what is in the `ZH` attribute but could be good to be explicit.

Reply:
good idea - I'll flesh out the ZH attribute